src/FEValues/face_values.jl

```diff
@@ -31,143 +31,57 @@ values of nodal functions, gradients and divergences of nodal functions etc. on
-    dNdx = fill(zero(dNdx_t) * T(NaN), n_func_basefuncs, max_n_qpoints, n_faces)
-    dNdξ = fill(zero(dNdξ_t) * T(NaN), n_func_basefuncs, max_n_qpoints, n_faces)
-
-    # Geometry interpolation
-    n_geom_basefuncs = getnbasefunctions(gip)
-    M    = fill(zero(T)      * T(NaN), n_geom_basefuncs, max_n_qpoints, n_faces)
-    dMdξ = fill(zero(dMdξ_t) * T(NaN), n_geom_basefuncs, max_n_qpoints, n_faces)
-
-    for face in 1:n_faces, (qp, ξ) in pairs(getpoints(qr, face))
-        for basefunc in 1:n_func_basefuncs
-            dNdξ[basefunc, qp, face], N[basefunc, qp, face] = shape_gradient_and_value(ip, ξ, basefunc)
-        end
-        for basefunc in 1:n_geom_basefuncs
-            dMdξ[basefunc, qp, face], M[basefunc, qp, face] = shape_gradient_and_value(gip, ξ, basefunc)
-        end
-    end
-
-    detJdV = fill(T(NaN), max_n_qpoints, n_faces)
-
-    FaceValues{IP, N_t, dNdx_t, dNdξ_t, T, dMdξ_t, QR, Normal_t, GIP}(N, dNdx, dNdξ, detJdV, normals, M, dMdξ, qr, ScalarWrapper(0), ip, gip)
-end
-
-# Common entry point that fills in the numeric type and geometric interpolation
-function FaceValues(qr::FaceQuadratureRule, ip::Interpolation,
-        gip::Interpolation = default_geometric_interpolation(ip))
-    return FaceValues(Float64, qr, ip, gip)
+function FaceValues(::Type{T}, fqr::FaceQuadratureRule, ip_fun::Interpolation, ip_geo::VectorizedInterpolation{sdim}=default_geometric_interpolation(ip_fun)) where {T,sdim} 
+    geo_values = [GeometryValues(T, ip_geo.ip, qr, RequiresHessian(ip_fun, ip_geo)) for qr in fqr.face_rules]
+    fun_values = [FunctionValues(T, ip_fun, qr, ip_geo) for qr in fqr.face_rules]
+    max_nquadpoints = maximum(qr->length(getweights(qr)), fqr.face_rules)
+    detJdV = fill(T(NaN), max_nquadpoints)
+    normals = fill(zero(Vec{sdim,T})*T(NaN), max_nquadpoints)
+    return FaceValues(fun_values, geo_values, fqr, detJdV, normals, ScalarWrapper(1))
 end
 
-# Common entry point that fills in the geometric interpolation
-function FaceValues(::Type{T}, qr::FaceQuadratureRule, ip::Interpolation) where {T}
-    return FaceValues(T, qr, ip, default_geometric_interpolation(ip))
+FaceValues(qr::FaceQuadratureRule, ip::Interpolation, args...) = FaceValues(Float64, qr, ip, args...)
+function FaceValues(::Type{T}, qr::FaceQuadratureRule, ip::Interpolation, ip_geo::ScalarInterpolation) where T
+    return FaceValues(T, qr, ip, VectorizedInterpolation(ip_geo))
 end
 
-# Common entry point that vectorizes an input scalar geometric interpolation
-function FaceValues(::Type{T}, qr::FaceQuadratureRule, ip::Interpolation, sgip::ScalarInterpolation) where {T}
-    return FaceValues(T, qr, ip, VectorizedInterpolation(sgip))
+function Base.copy(fv::FaceValues)
+    fun_values = map(copy, fv.fun_values)
+    geo_values = map(copy, fv.geo_values)
+    return FaceValues(fun_values, geo_values, copy(fv.qr), copy(fv.detJdV), copy(fv.normals), copy(fv.current_face))
```

Comment:
So, for me to understand this, here copy acts like a deepcopy?

Reply:
Yes, but @fredrik explained to me that one shouldn't use `deepcopy` actually, but define `copy` to work this way when required. 
(Note that this is no change from before, it is just that the automatic method doesn't work so well due to the nesting, see https://github.com/Ferrite-FEM/Ferrite.jl/blob/e9ae940a3b6965d78bfd1ae03827e7584bfddac4/src/FEValues/common_values.jl#L276 Actually, the [landau example](https://github.com/Ferrite-FEM/Ferrite.jl/blob/e9ae940a3b6965d78bfd1ae03827e7584bfddac4/docs/src/literate-gallery/landau.jl#L111) relies on this.